eventsourcing/src/main/java/org/axonframework/eventsourcing/MultiStreamableMessageSource.java

```diff
@@ -0,0 +1,384 @@
+         * @param trackingToken  the {@link MultiSourceTrackingToken} containing the positions to open the stream.
+         */
+        public MultiSourceMessageStream(Map<String, StreamableMessageSource<TrackedEventMessage<?>>> messageSources,
+                                        MultiSourceTrackingToken trackingToken) {
+            this.messageSources = new HashMap<>();
+            messageSources.forEach((k, v) -> this.messageSources
+                    .put(k, v.openStream(trackingToken.getTokenForStream(k))));
+        }
+
+        /**
+         * @return true if one of the streams has a message available to be processed. Otherwise false.
+         */
+        @Override
+        public boolean hasNextAvailable() {
+            for (Map.Entry<String, BlockingStream<TrackedEventMessage<?>>> singleMessageSource : messageSources
+                    .entrySet()) {
+                if (singleMessageSource.getValue().hasNextAvailable()) {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * Peeks each stream to check if a message is available. If more than one stream has a message it returns the
+         * message chosen using the comparator.
+         *
+         * @return message chosen using the comparator.
+         */
+        @Override
+        public Optional<TrackedEventMessage<?>> peek() {
+            if (peekedMessage == null) {
+                peekedMessage = doConsumeNext();
+            }
+
+            return Optional.ofNullable(peekedMessage);
+        }
+
+        private TrackedEventMessage<?> doConsumeNext() {
+            HashMap<String, TrackedSourcedEventMessage> candidateMessagesToReturn = new HashMap<>();
+
+            peekForMessages(candidateMessagesToReturn);
+
+            //select message to return from candidates
+            final Optional<Map.Entry<String, TrackedSourcedEventMessage>> chosenMessage =
+                    candidateMessagesToReturn.entrySet().stream().min((m1, m2) -> comparator
+                            .compare(m1.getValue(), m2.getValue()));
+
+            // Ensure the chosen message is actually consumed from the stream
+            return chosenMessage.map(e -> {
+                String streamId = e.getKey();
+                TrackedEventMessage<?> message = e.getValue();
+                try {
+                    return new GenericTrackedEventMessage<>(trackingToken.advancedTo(streamId, message.trackingToken()),
+                                                            messageSources.get(streamId).nextAvailable());
+                } catch (InterruptedException ex) {
+                    logger.warn("Thread Interrupted whilst consuming next message", ex);
+                }
+                return null;
+            }).orElse(null);
+        }
+
+        /**
+         * @param timeout the maximum number of time units to wait for messages to become available. This time
```

Comment:
I think this method also deserves a JavaDoc body. But, if it's identical to the method which is being overridden, I think we can drop this JavaDoc entirely as it will be inherited automatically.

Reply:
I've added a body since I think its behaviour requires some explaining